### PR TITLE
Share logic for saving page/cover

### PR DESCRIPTION
**Pull request by ghostbear**

### CODE REVIEW

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaPresenter.kt [hunk 1]
```diff
@@ -110,6 +108,8 @@ class MangaPresenter(
 
     private val loggedServices by lazy { trackManager.services.filter { it.isLogged } }
 
+    private val imageSaver by lazy { ImageSaver(preferences.context) }
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaPresenter.kt:111]: Could maybe add it to `AppModule` and inject it instead?

File: app/src/main/java/eu/kanade/tachiyomi/data/saver/ImageSaver.kt [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        return destFile.getUriCompat(context)
+    }
+}
+
+sealed class Image(
+    open val name: String,
+    open val location: Location
+) {
+    data class Cover(
+        val bitmap: Bitmap,
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/data/saver/ImageSaver.kt:94]: I _think_ Coil 2.x might allow us to get non-bitmap versions of the images, but we could look into that separately.

File: app/src/main/java/eu/kanade/tachiyomi/data/saver/ImageSaver.kt [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    object Cache : Location()
+
+    fun directory(context: Context): File {
+        return when (this) {
+            Cache -> context.cacheImageDir
+            is Pictures -> {
+                val file = File(
+                    Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES),
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/data/saver/ImageSaver.kt:136]: Was part of the motivation of this PR to remove usage of deprecated APIs?

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaPresenter.kt [hunk 4]
```diff
@@ -338,44 +338,13 @@ class MangaPresenter(
...
 
     /**
-     * Save manga cover Bitmap to temporary share directory.
+     * Save manga cover Bitmap to picture or temporary share directory.
      *
-     * @param context for the temporary share directory
-     * @param coverBitmap the cover to save (as Bitmap)
-     * @return cover File in temporary share directory
+     * @param image the image with specified location
+     * @return flow Flow which emits the Uri which specifies where the image is saved when
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaPresenter.kt:344]: Why a flow and not just a suspension function?

File: app/src/main/java/eu/kanade/tachiyomi/data/saver/ImageSaver.kt (hunk 3 above)
Comment (ghostbear) [reply app/src/main/java/eu/kanade/tachiyomi/data/saver/ImageSaver.kt:136]: Yeah, but they are still used as MediaStore before API level 29 doesn't have all functions needed to remove them entirely

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaPresenter.kt (hunk 4 above)
Comment (arkon) [reply app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaPresenter.kt:345]: Doesn't it seem awkward to keep the flow around until the scope's cancelled despite it not emitting multiple items?

**Pull request merged**